include/Materials/DruckerPrager.h

```
@@ -0,0 +1,307 @@
+      Eigen::Matrix<double, 3, 3> dee(dstrain);
+      Eigen::Matrix<double, 3, 3> dep;
+      dep.setZero();
+
+      auto f = m_A1 - m_A2 * zt;
+      auto g = st / std::sqrt(2.0) - f;
+
+      if (g > 1.0e-4)
+      {
+        // plastic
+        bool converged = false;
+        
+        // Apply oblique return to put stress on yield surface
+        
+        // Newton iterations to find magnitude of projection from the trial stress state to the yield
+        // surface.
+        for (auto iter=0; iter<25; iter++)
+        {
+          // Yield normal N and flow direction M
+          auto Nz = m_A2;
+          auto Ns = 1.0 / std::sqrt(2.0);
+    
+          // Flow direction
+          auto Mz = m_A2G;
+          auto Ms = Ns;
+    
+          Eigen::Matrix<double, 3, 3> N;
+          Eigen::Matrix<double, 3, 3> M;
+          for (auto i=0; i<3; i++)
+          {
+            for (auto j=0; j<3; j++)
+            {
+              N(i, j) = Nz * Ez(i, j) + Ns * Es(i, j);
+              M(i, j) = Nz * Ez(i, j) + Ns * Es(i, j);
+            }
+          }
+    
+          // Components of A tensor
+          auto Az = m_bulk_modulus * Mz;
+          auto As = 2.0 * m_shear_modulus * Ms;
+    
+          // Scaled components of return direction tensor
+          auto fac = std::sqrt((zt * zt + st * st) / (Mz * Mz + Ms * Ms)) / m_bulk_modulus;
+          auto Pz = fac * Az;
+          auto Ps = fac * As;
+    
+          // Projection "magnitude"
+          auto beta = -g / (Nz * Pz + Ns * Ps);
+    
+          // Check for convergence
+          if(std::abs(beta) < 1.0e-10)
+          {
+            converged = true;
+            break;
+          }
+    
+          // Improved estimates for the test stress
+          zt += beta * Pz;
+          st += beta * Ps;
+
+          f = m_A1 - m_A2 * zt;
+          g = st / std::sqrt(2.0) - f;
+        }
+        
+        if (!converged)
+          throw std::logic_error("Drucker-Prager Newton iterations failed");
```

Maybe Kokkos::Abort instead of std::logic_error.